Documentation/botr/readytorun-format.md

```diff
@@ -4,10 +4,70 @@ ReadyToRun File Format
+
+This document describes ReadyToRun format 3.1 implemented in CoreCLR as of June 2019 and not yet implemented proposed extensions 3.2 for the support of composite R2R file format. **Composite R2R file format** has basically the same structure as the traditional R2R file format defined in earlier revisions except that the output file represents a larger number of input MSIL assemblies compiled together as a logical unit.
+
+## Traditional vs. composite R2R image
+
+The logical goal of a composite R2R image is to have a single PE R2R binary represent a semantic union of multiple input MSIL assemblies compiled as a unit by the R2R compiler to produce the composite output.
+
+Some motivations for producing composite R2R images include:
+
+* Better performance: compiling multiple MSIL assemblies into a single output R2R executable creates an even tighter coupling than the large version bubble using separate files and the compiler and runtime can be more efficient (e.g. the compiler may elide MVID checks and type layout checks within the "tight" version bubble).
+* Easier deployment: desire to avoid having to deal with potentially hundreds or thousands of assemblies.
+* Side-by-side deployment: a self-contained app doesn't need to care about the location of its dependencies, their proper versions, about sharing them with other apps (including other versions of the same app).
+
+Semantic requirements for the composite R2R format flow from the combination of the basic design principle that the R2R file is a code cache to save jitting time and from the fact that the tighter coupling can be utilized to improve codegen quality.
+
+### R2R file is a code cache
+
+A corollary of this basic principle is that the composite file must maintain full functionality of single-input R2R files w.r.t. things like reflection, dynamic or generics. To achieve this goal the composite R2R file must store all the original MSIL metadata blobs for all input assemblies in the output R2R file. This is a substantial deviation from the previous format that only had one COR header addressed by the COM data directory entry in the COFF header.
+
+To achieve this goal, we propose using two complementary strategies:
+
+* In the composite R2R file, there must be a new table of COR headers and metadata blobs representing the MSIL metadata from all the input assemblies. The table must be indexable by simple assembly name for fast lookup. 
+
+* We propose slightly variating the purpose of the "main" COR header and metadata blob: in composite R2R files, where a "singleton" ECMA-335 metadata blob doesn't make much sense, we'd reuse the "main" metadata for the manifest metadata.
```

Comment:
I do not see what we are getting by making the composite format look like a regular IL image and placing the manifest metadata in place of regular metadata. It is just going to confuse every tool out there when looking at these files.

Would it be better to make the composite format to look like unmanaged PE file, with import pointing the the R2R header? 

It would get us onto a path to have an option to make R2R images look like native Unix .so files.

Reply:
I like this idea. I have updated the documentation in this sense.